doc/error-handling.md

```diff
@@ -0,0 +1,324 @@
+
+
+## Error Handling in posh
+Error logging shall be done by the logger only, no calls to std::cerr or similar should be performed.
+
+All the methods presented (cxx::expected, Expects and Ensures and the error handler) can be used in posh. The appropriate way depends on the type of error scenario (cf. the respective sections for examples). The error handler should be considered the last option.
+
+## Error Handling in utils
+Error logging is currently done by calls to cerr. In the future those might be redirected to the logger.
+
+The error handler cannot be used in utils. 
+
+Whether it is appropriate to use cxx::expected even if STL compatibility is broken by doing so depends on the circumstances and needs to be decided on a case-by-case basis. If the function has no STL counterpart cxx::expected can be used freely to communicate potential failure to the caller.
+
+It should be noted that since currently Expects and Ensures are active at release mode, prolific usage of these will incur a runtime cost. Since this is likely to change in the future, it is still advised to use them to document the developers intentions.
+
+## Interface for 3rd Party Code
+
+Error handler as well as logger shall be able to use or redirect to 3rd party error handling or logging libraries in the future. Currently this is neither fully supported nor used. The error handler as a callback function which can in principle used to call 3rd party code.
+
+# Usage
+
+## Logger
+The logger can be used similar to the streams in the C++ standard API.
+To select the log level, the corresponding logger has to be used, e.g. LogErr, LogWarn etc.
+
+```
+LogWarn() << "log message " << someValue << "log message continued";
+```
+A line break is inserted implicitly at the end (after "log message continued" in the example).
+
+## Error Handler
+The most general use case is the following
+```
+if(noError()) {
+    //handle the regular case
+} else {
+    auto callback = []() { //some error handling callback};
+    errorHandler(Error::kSOME_ERROR_CODE, callback, ErrorLevel::SEVERE);
+}
+```
+
+If no callback or error level are specified, the error level is assumed to be FATAL by default.
+```
+errorHandler(Error::kSOME_ERROR_CODE)
+```
+
+If no callback but an error level is desired, a nullptr has to be provided for the callback.
+```
+errorHandler(Error::kSOME_ERROR_CODE, nullptr, ErrorLevel::MODERATE);
+```
+
+## Expects and Ensures
+
+Assume func is part of an inner API and not supposed to be called with a nullptr. We may have used a reference here, this is just for illustration.
+In addition the value pointed to is assumed to be in the range (-1024, 1024). While we could check this everytime, this can be avoided if we specify that the caller is responsible to ensure that these conditions hold.
+
+```
+int myAlgorithm(int* ptr) {
```

Comment:
Shouldn't we use fixed width integer types in these kind of examples as well?

Reply:
Can do, sure. But it is not the focus of the example of course.